### ISSUE TITLE
Deserialization error with multiple resources

### ISSUE THREAD

**Issue by winteler** [leptos-rs/leptos#2188]
Hi Greg,

I think I found another bug with resources where leptos mixes up the return types of two resources, leading to a deserialization error. This might be similar (or even the same issue as) https://github.com/leptos-rs/leptos/issues/2063 but appearing under different circumstances and after trying a few days I wasn't able to come up with a workaround.

**Leptos Dependencies**

```toml
leptos = { version = "0.5", features = ["nightly"] }
leptos_axum = { version = "0.5", optional = true }
leptos_meta = { version = "0.5", features = ["nightly"] }
leptos_router = { version = "0.5", features = ["nightly"] }
```

**To Reproduce**

Here is an example to reproduce the issue:

```
use leptos::*;
use leptos_meta::*;
use leptos_router::*;

use crate::error_template::{AppError, ErrorTemplate};

#[server]
pub async fn empty_data() -> Result<(), ServerFnError> {
    Ok(())
}

#[server]
pub async fn int_data() -> Result<i32, ServerFnError> {
    Ok(1)
}

#[server]
pub async fn char_data() -> Result<char, ServerFnError> {
    Ok('c')
}

#[server]
pub async fn string_data() -> Result<String, ServerFnError> {
    Ok(String::from("Test"))
}

#[server]
pub async fn bool_data() -> Result<bool, ServerFnError> {
    Ok(true)
}

#[component]
pub fn App() -> impl IntoView {
    // Provides context that manages stylesheets, titles, meta tags, etc.
    provide_meta_context();
    view! {
        // injects a stylesheet into the document <head>
        // id=leptos means cargo-leptos will hot-reload this stylesheet
        <Stylesheet id="leptos" href="/pkg/leptos-reproduce-bugs.css"/>

        // sets the document title
        <Title text="Welcome to [[ProjectName]]"/>

        // content for this welcome page
        <Router fallback=|| {
            let mut outside_errors = Errors::default();
            outside_errors.insert_with_default_key(AppError::NotFound);
            view! {
                <ErrorTemplate outside_errors/>
            }
            .into_view()
        }>
            <main>
                <div>
                    <div>
                        <Routes>
                            <Route path="/" view=ContainerHeader>
                                <Route path="/" view=ContainerContent/>
                                <Route path="/object" view=Object/>
                            </Route>
                        </Routes>
                    </div>
                </div>
            </main>
        </Router>
    }
}

#[component]
pub fn ContainerHeader() -> impl IntoView {
    let int_data = Resource::once(int_data);
    view! {
        <Suspense>
            <pre>{
                move || {
                    int_data.with(|result| match result {
                        Some(Ok(data)) => {
                            view! {
                                <div>{format!("{:?}", data)}</div>
                                <Outlet/>
                            }.into_view()
                        },
                        _ => {
                            view! { <Outlet/> }.into_view()
                        }
                    })
                }
            }</pre>
        </Suspense>
    }
}

#[component]
pub fn ContainerContent() -> impl IntoView {
    let int_data = Resource::once(char_data);
    view! {
        <Suspense>
            <pre>{
                move || {
                    view! {
                        <div>{format!("{:?}", int_data.get())}</div>
                        <a href="/object">"Object 1"</a>
                    }
                }
            }</pre>
        </Suspense>
    }
}

#[component]
pub fn Object() -> impl IntoView {
    let int_data = Resource::once(string_data);
    view! {
        <Suspense>
            <pre>{
                move || {
                    view! {
                        <div>{format!("{:?}", int_data.get())}</div>
                    }
                }
            }</pre>
        </Suspense>
        <Detail/>
    }
}

#[component]
pub fn Detail() -> impl IntoView {
    let char_data = Resource::once(bool_data);
    view! {
        <Suspense>
            <pre>{
                move || {
                    view! {
                        <div>{format!("{:?}", char_data.get())}</div>
                    }
                }
            }</pre>
        </Suspense>
    }
}
```

The error occurs when clicking the 'Object 1' link, giving the error:

could not deserialize Resource<core::result::Result<bool, server_fn::error::ServerFnError>> JSON for ResourceId(3v1): Deserialize(Error("invalid type: string \"c\", expected a boolean", line: 1, column: 9))

Would you have a workaround idea for this issue? It's not a huge issue but it would be nice to not face this error when developing/testing :)

Thanks for your help!

**Comment by gbj** [comment]
If you open View Source you'll see that the `'c'` server function resource is being created twice. These kinds of nested Suspense/Outlet patterns naturally create async data loading waterfalls that are bad -- in your example, the `<Outlet/>` is being created whether the resource you're waiting on is loaded or not, and so the resource is created twice, so the server function is called twice, etc.

For this particular example the fix is as simple as moving the Outlet outside the Suspense entirely:
```rust
#[component]
pub fn ContainerHeader() -> impl IntoView {
    let int_data = Resource::once(int_data);
    view! {
        <Suspense>
            <pre>
                {move || {
                    int_data.with(|result| match result {
                        Some(Ok(data)) => {
                            Some(view! {
                                <div>{format!("{:?}", data)}</div>
                            })
                        },
                        _ => None
                    })
                }}
            </pre>
        </Suspense>
        <Outlet/>
    }
}
```

However this is also a bug, in that the reactive system is generating overlapping resource IDs.

**Comment by winteler** [comment]
Thanks a lot for the (very!) quick answer :) this indeed solved the issue.

I tried to have a look at the Page Source to see on my side the resource being created twice but I have to admit I was a bit lost in there '^^ would you have some more indications to find where in the code the resource is created?

Thanks again for your help!

**Comment by gbj** [comment]
Just tested and this is fixed on 0.7, so I will close this issue.
